accepted/future-releases/language-versioning/package-config-file-v2.md

```diff
@@ -0,0 +1,193 @@
+
+Pub has stronger restrictions on its accepted package names, but this file will also be used in settings where packages are not supplied by Pub.
+
+### Finding the File
+
+When compiling a Dart script file, the Dart tool will look for the necessary information starting in the directory of the script file. It will first check for a `.dart_tool/package_config.json` file, then for a `.packages` file, the it will repeat this through the parent directories until it one such file or it reaches the file system root.
+
+The analyzer needs to cleverly determine which configuration file applies to which file, based on the location of scripts, which scripts may transitively import a file, and understanding of the structure of Pub packages. It should be able to use the same algorithm as now, just that where it currently looks for a `.packages` in a directory, it will have to for `.dart_tool/package_config.json` first.
+
+### Specifying the File
+
+Tools currently allow you to specify the `.packages` file with a command line argument:
+
+```
+--packages=someFile
+-p someFile
+```
+
+This approach will work with the JSON configuration file as well. The tool must scan the file to detect whether it is a JSON file or an original `.packages` file. It is considered a JSON file if the first non-whitespace (space, tab, carriage return, linefeed) character is a `{`. The JSON file must start this way, and the existing format cannot do so; it must start with either a comment, `#`, or a non-empty package name, and package names cannot contain a `{` character.
+
+## Possible Extensions
+
+We can consider already adding more features to the format. These features should be ones that are of relevance to compilers reading source files, since that is the goal of the file.
+
+Any such extensions would need to be added by the generator, so in practice they must come from `pubspec.yaml` to begin with. If the Pub tool wants to support these, it is possible. 
+
+### Experiments
+
+Maybe you can enable experiments for a specific package by adding, for example:
+
+```json
+"experiments": ["non-nullable"]
+```
+
+to a single package entry, or add it in the outer object to enable it for all compatible packages.
+
+### Environment Variables
+
+Maybe you can hard-code some environment variables by adding, for example:
+
+```json
+"environment": {
+  "myFlag": "true",
+  "logLevel": "verbose"
+}
+```
+
+in the outer object. These would be available for `String.fromEnvironment("logLevel", defaultValue: "default")` or `bool.fromEnvironment("myFlag")`
+
+## Summary
+
+A file named `package_config.json`, stored in `.dart_tool`, will coexist with `.packages` until `.packages` can be phased out.
+
+The format will look like:
+
+```json
+{
+  "apiVersion": 2,
+  "packages": [
+    {
+      "name": "myPackage",
+      "rootUri": ".",
```

Comment:
This needs to be `../` if we assume that `package_config.json` eixsts in `./dart_tool/package_config.json` and according to this spec:
> The URI reference is resolved relative to the (likely file:) URI identifying the package_config.json file.